### PR TITLE
[JENKINS-60995] Implement a specific UpstreamCause class containing the nodeId that triggered the build

### PR DESCRIPTION
**Pull request by Vlatombe**
This is part of a bigger change for blueocean in order to fix downstream links.
* (master) https://github.com/jenkinsci/blueocean-plugin/pull/2187
* (backport on 1.24.x) https://github.com/jenkinsci/blueocean-plugin/pull/2188

- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [X] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Review by bitwiseman — APPROVED** [review]

**Comment by bitwiseman** [comment]
Oops didn't need rerunning.  Using incremental in downstream PR: 2.14-rc430.69345dedbb03.

**Comment by Vlatombe** [comment]
@bitwiseman I wrote a better test, for a scenario that was typically a problem when relying on the `BuildUpstreamNodeAction` to obtain a backlink to the upstream job.

File: src/test/java/org/jenkinsci/plugins/workflow/support/steps/build/BuildTriggerStepTest.java [hunk 1]
```diff
@@ -115,6 +124,54 @@
...
+        for (WorkflowRun upstreamRun : upstream.getBuilds()) {
+            j.waitForCompletion(upstreamRun);
+            j.assertBuildStatus(Result.SUCCESS, upstreamRun);
+        }
+        // Wait for all downstream builds to complete
+        j.waitUntilNoActivity();
+
+        List<BuildUpstreamCause> upstreamCauses = new ArrayList<>();
+        for (Run<FreeStyleProject, FreeStyleBuild> downstreamRun : downstream.getBuilds()) {
+            upstreamCauses.addAll(downstreamRun.getCauses().stream().filter(BuildUpstreamCause.class::isInstance).map(BuildUpstreamCause.class::cast).collect(Collectors.toList()));
```
Comment (bitwiseman) [inline src/test/java/org/jenkinsci/plugins/workflow/support/steps/build/BuildTriggerStepTest.java:148]: Readability nitpick:
```suggestion
            upstreamCauses.addAll(downstreamRun.getCauses().stream()
                .filter(BuildUpstreamCause.class::isInstance)
                .map(BuildUpstreamCause.class::cast)
                .collect(Collectors.toList()));
```

**Review by bitwiseman — APPROVED** [review]

**Review by olamy — APPROVED** [review]

**Comment by saper** [comment]
Just FYI: We were depending (incorrectly) on upstream causes being instances of `hudson.model.Cause$UpstreamCause`, now they are instances of `org.jenkinsci.plugins.workflow.support.steps.build.BuildUpstreamCause`.

We were doing it wrong, but I hope this didn't break something somewhere else.

**Pull request merged**